### ISSUE TITLE
Speeding up travis builds -- ideas?

### ISSUE THREAD

**Issue by sethladd** [dart-lang/dartdoc#538]
Travis builds are pretty slow. What can we do to speed them up?

Right now, we're generating the docs for:

* test_package
* dartdoc itself
* the SDK

Do we need to generate the docs for all of these, on each build?

**Comment by sethladd** [comment]
cc @keertip @devoncarew 

Here's an idea:

Create two Travis configs. A fast "early warning" build and a slow "all the things" build.

The fast version runs all the tests, gets code coverage, and builds docs for test_package.

The slow version generates the docs for the core SDK.

We can merge if the fast version succeeds. And we'll still be notified if the slow build fails.

Can we do this?

**Comment by sethladd** [comment]
cc @a14n who knows about Travis builds, too, for his thoughts on running two different travis.sh scripts for a single commit. Is that OK?

**Comment by devoncarew** [comment]
I think you can shard travis builds based on env variables, and then your build script can check for the value or presence of different env. variables.

We're already sharing on dev or stable sdks, so adding an BUILD_SDK var would mean we have 4 builds going per PR (2 sdks x 2 SDK flags)

**Comment by sethladd** [comment]
See also http://docs.travis-ci.com/user/speeding-up-the-build/

**Comment by sethladd** [comment]
PR https://github.com/dart-lang/dartdoc/pull/541

**Comment by sethladd** [comment]
Neat!

![screen shot 2015-05-29 at 1 14 47 pm](https://cloud.githubusercontent.com/assets/5479/7891838/b97e5694-0604-11e5-9f16-d4bae62bda0f.png)

**Comment by sethladd** [comment]
I also turned on concurrent builds to 4. Not sure why it was at zero, or what the trade-offs are.

**Comment by sethladd** [comment]
Some good news: we have matrix builds set up. However, the checks on the PR page on github do not turn green until both GEN_SDK_DOCS runs (true, and false) are complete.

**Comment by sethladd** [comment]
Maybe @devoncarew or @a14n know how to configure GitHub to show more checks for a PR (one check for each matrix configuration...)

**Comment by a14n** [comment]
You can try to add:

```yaml
matrix:
  fast_finish: true
```

See [Fast finishing](http://docs.travis-ci.com/user/build-configuration/#Fast-finishing).

**Comment by a14n** [comment]
Never used before but [Rows That are Allowed To Fail](http://docs.travis-ci.com/user/build-configuration/#Rows-That-are-Allowed-To-Fail) seems to be what you are looking for to have the workflow you described

>We can merge if the fast version succeeds. And we'll still be notified if the slow build fails.

**Comment by sethladd** [comment]
awesome, I'll look into that. Thanks!

How does GitHub get two checks from Travis? Not sure how that's setup... how did it know to check for each Dart SDK version?

**Comment by sethladd** [comment]
A little faster now:

![screen shot 2015-05-29 at 2 16 49 pm](https://cloud.githubusercontent.com/assets/5479/7892935/711cd4a8-060d-11e5-83be-6fd95fce7fa2.png)

**Comment by sethladd** [comment]
PR to use fast_fail: https://github.com/dart-lang/dartdoc/pull/543

Still, am curious where the names travis-ci/pr and travis-ci/push come from...

**Comment by a14n** [comment]
see https://travis-ci.org/dart-lang/dartdoc/settings

**Comment by sethladd** [comment]
Oh! interesting... maybe we're doing too much work. If I push to a branch, and then make a pull request, that's a double build.

I think I only care to build when I push to master, or when a PR is submitted. Does that make sense?

**Comment by a14n** [comment]
Last clue for tonight acknowledging the times on the builds : setup a cache directory for pub

```yaml
cache:
  directories:
  - ".pub-cache/"
```

See http://docs.travis-ci.com/user/caching/#Arbitrary-directories

**Comment by sethladd** [comment]
Each build now runs in parallel, and a single build takes 1min 30sec. We're saving a minute now. Also, we're no longer building twice on every push. (only pushes to master, and on every pull request)

![screen shot 2015-05-29 at 2 45 53 pm](https://cloud.githubusercontent.com/assets/5479/7893354/c98e014e-0611-11e5-87d6-d001246a4660.png)

**Comment by sethladd** [comment]
another PR https://github.com/dart-lang/dartdoc/pull/545

**Comment by sethladd** [comment]
Related: only run the tests once https://github.com/dart-lang/dartdoc/issues/539

**Comment by alan-knight** [comment]
Not really related, but is there way to get the Travis builds to run at all? I have a trivial project where I was setting it up, but it seems to take it on the order of a day to get around to running tests at all. It would be faster on the buildbots.

**Comment by sethladd** [comment]
let's chat about this in another thread, but tldr: your builds should be starting in under a minute, after a pull request or push is sent.

**Comment by devoncarew** [comment]
I've noticed that the containerized builds start up faster (`sudo: false`).

**Comment by devoncarew** [comment]
https://github.com/dart-lang/dartdoc/blob/master/.travis.yml#L2

**Comment by sethladd** [comment]
Here's another idea: we can parallelize much of the build by using Travis' matrix builds. Many of the dartdoc tasks can be run in parallel.

**Comment by a14n** [comment]
Looking at the times in the builds, it seems that most of the time is consume by the execution of tool/travis.sh. I don't think we can do further optimization on the travis side (once #546 merged)

**Comment by sethladd** [comment]
Builds times are now around 1min 30sec

![screen shot 2015-06-01 at 9 55 58 am](https://cloud.githubusercontent.com/assets/5479/7918206/852d41ec-0844-11e5-8858-6104e220ec04.png)

**Comment by sethladd** [comment]
Closing for now. We might be able to shard the builds even more, but for now, moving on.

Thanks for all the help!

**Comment by a14n** [comment]
The build time seems to change depending on the datetime you trigger the build. For instance see https://travis-ci.org/a14n/dartdoc/builds/64707303 that took less than 1 min!

Another way to optimize the build time is to work during the weekend :D
